Migrate "old" conditions

Cyperghost opened the pull request:
Ref https://github.com/WoltLab/WCF/pull/6304
See https://github.com/WoltLab/WCF/issues/6310

### TODO

- [ ] Change branch to `6.3-condition`

Cyperghost committed:
Add IMigrateConditionType interface and implement migration methods in user group condition types

Cyperghost committed:
Implement migration support for user condition types

Cyperghost committed:
Add migration support for user group assignment conditions

Cyperghost committed:
Use a unique structure for the object types for migration.

dtdesign commented on wcfsetup/install/files/acp/database/update_com.woltlab.wcf_6.3_step1.php:
```diff
@@ -8,12 +8,14 @@
     PartialDatabaseTable::create('wcf1_user_group_assignment')
         ->columns([
             MediumtextDatabaseTableColumn::create('conditions'),
+            DefaultFalseBooleanDatabaseTableColumn::create('needMigration'),
```
I would prefer something like `isLegacy` which is more in line with our general naming schema.

dtdesign commented on wcfsetup/install/files/lib/system/user/group/assignment/command/UserGroupAssignmentMigrateCondition.class.php:
```diff
@@ -0,0 +1,42 @@
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.3
+ */
+final class UserGroupAssignmentMigrateCondition
```
This completely violates the naming schema for commands.

dtdesign commented on wcfsetup/install/files/lib/system/cache/eager/UserGroupAssignmentCache.class.php:
```diff
@@ -0,0 +1,50 @@
+        if ($migrationDone) {
+            // Reload the list to ensure that no disabled assignments are included
+            return $this->getUserGroupAssignments()->getObjects();
+        } else {
```
This is a very odd place to perform a migration, business logic doesn’t belong here.

Migrations are only ever relevant after a package update has completed because that is the only ingress point for new implementations.

dtdesign commented on wcfsetup/install/files/lib/system/condition/ConditionHandler.class.php:
```diff
@@ -172,4 +176,82 @@ public function getConditionsWithFilter(AbstractConditionProvider $provider, arr
+
+        $conditionBuilder = new PreparedStatementConditionBuilder();
+        $conditionBuilder->add('objectTypeID IN (?)', [
+            \array_map(static fn (ObjectType $objectType): int => $objectType->objectTypeID, $objectTypes),
```
```suggestion
            \array_column($objectTypes, 'objectTypeID'),
```

dtdesign commented on wcfsetup/install/files/lib/system/condition/ConditionHandler.class.php:
```diff
@@ -172,4 +176,82 @@ public function getConditionsWithFilter(AbstractConditionProvider $provider, arr
+                    break;
+                }
+            }
+        }
```
```suggestion
        }
        \unset($condition);
```

Please always unset variables bound by referenc. This can easily lead to unintended behavior later on because the reference escapes the scope of the `foreach`.

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/ConditionMigration.class.php:
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace wcf\system\condition\provider;
```
What is the rationale of placing this in the `provider` sub namespace?

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/ConditionMigration.class.php:
```diff
@@ -0,0 +1,38 @@
+     * @param array{identifier: string, value: mixed}[] $previousConditionData
+     * @param array{identifier: string, value: mixed}[] $migratedConditionData
+     */
+    public static function for(array $previousConditionData, array $migratedConditionData): self
```
`fromData()` perhaps is a more fitting name.

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/ConditionMigration.class.php:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Creates a new ConditionMigration instance for empty data.
+     */
+    public static function forNoConditions(): self
```
`withoutData()` makes it less susceptible for an incorrect auto completion and is in line with other constructs like the `dialogFactory()` in TypeScript.

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/ConditionMigration.class.php:
```diff
@@ -0,0 +1,38 @@
+final class ConditionMigration
+{
+    private function __construct(
+        public readonly bool $isFullMigrated,
```
I don’t understand the point of this variable (besides it missing a `y`), isn’t the presence of conditions sufficient?

Cyperghost replied:
No, it is possible that only 99 out of 100 different conditions were migrated. This should not be considered a full migration. This can occur if, for example, a condition type was added by a third party plugin and does not (yet) exist.
The previous user group assignment must therefore be deactivated, otherwise users could be assigned under different circumstances than before.

However, if we only allow/enable the migration of data if all data has been transferred correctly, it _may happen_ that the admin has to re-enter all data for a user group assignment because a plugin is no longer supported.

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/ConditionMigration.class.php:
```diff
@@ -0,0 +1,38 @@
+    private function __construct(
+        public readonly bool $isFullMigrated,
+        /** @var array{identifier: string, value: mixed}[] */
+        public readonly array $conditions = [],
```
Please don’t use default values together with fixed construct methods. Instead have the methods explicitly define all parameters.

dtdesign commented on wcfsetup/install/files/lib/system/condition/provider/UserConditionProvider.class.php:
```diff
@@ -35,25 +35,25 @@ final class UserConditionProvider extends AbstractConditionProvider
         $this->addConditions([
-            new class("username", "username") extends AbstractUserStringConditionType {},
-            new class("email", "email") extends AbstractUserStringConditionType {},
+            new class("username", "username", "username", 'com.woltlab.wcf.user.username') extends AbstractUserStringConditionType {},
```
This (among others) crosses the threshold for named parameters. Maybe even wrap the parameters to make it an easy to parse `key: "value"` list.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/IMigrateConditionType.class.php:
```diff
@@ -0,0 +1,23 @@
+     */
+    public function migrateConditionData(array &$conditionData): array;
+
+    public function canMigrateConditionData(string $objectType): bool;
```
This lacks a description that helps the developer understand what this method is supposed to validate. When is it reasonable not to support a migration?

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/IMigrateConditionType.class.php:
```diff
@@ -0,0 +1,23 @@
+ */
+interface IMigrateConditionType
+{
+    /**
+     * Converts the old condition structure to the new one. All migrated values must be removed from the `$conditionData` array.
+     *
+     * @param array<string, mixed> $conditionData
+     *
+     * @return array{identifier: string, value: mixed}[]
+     */
+    public function migrateConditionData(array &$conditionData): array;
```
This entire methods makes me feel uneasy:

1. References in parameters should be avoided like the plague because PHP does not require the call-side to mark references which makes these things quite obscure.
2. If the implementation is supposed to strip data from `$conditionData` then what is expected to be returned from it? The migrated data? And if yes, why bother removing it from `$conditionData`?
3. `array{…}[]` is an awful way to document the type, please use `[]` only for simple expressions like `int[]`. This should be `list<array{…}>` or something like that.

Cyperghost replied:
It is not easy to determine whether all data has been migrated correctly and completely without using a reference.
For example, [`UserGroupCondition`](https://github.com/WoltLab/WCF/blob/6.1/wcfsetup/install/files/lib/system/condition/UserGroupCondition.class.php) now consists of two classes:  `UserNotInGroupConditionType` and `UserInGroupConditionType`.
Therefore, I cannot tell whether, for example, after migrating `UserInGroupConditionType`, further steps are necessary for the object type `com.woltlab.wcf.userGroup` or not.

If I were to remove the data `$conditionData[‘com.woltlab.wcf.userGroup’]` after the first call to `migrateConditionData()`, only `UserInGroupConditionType::migrateConditionData()` would be executed, but not `UserNotInGroupConditionType::migrateConditionData()`.

Therefore, the class must remove all entries it can process so that `ConditionHandler::migrateConditionData()` can recognize whether everything has been transferred or not.

dtdesign replied:
Understood. In that case please expand the existing documentation to explain this which also serves as a guideline for developers to properly use this API.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserBooleanConditionType.class.php:
```diff
@@ -65,4 +68,28 @@ public function matches(object $object): bool
+    #[\Override]
+    public function canMigrateConditionData(string $objectType): bool
+    {
+        return $this->migrateConditionObjectType === null || $this->migrateConditionObjectType === $objectType;
```
Why does `$this->migrateConditionObjectType === null` qualify this type to perform a migration? What is that property even supposed to do, it is not used in `migrateConditionData()` at all?

Cyperghost replied:
`null` should of course not allow migration, as is the case everywhere else.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserBooleanConditionType.class.php:
```diff
@@ -65,4 +68,28 @@ public function matches(object $object): bool
+            return [];
+        }
+
+        $value = $conditionData[$this->migrateKeyName];
```
This can be simplified as `$value = $conditionData[$this->migrateKeyName] ?? null;` and then to return when `$value` is `null`.

This works because `null` when used as an array key is interpreted as an empty string and `isset()` performs a null-check already, thus `null` is never a valid value.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserIntegerConditionType.class.php:
```diff
@@ -84,8 +86,39 @@ public function matches(object $object): bool
+    #[\Override]
+    public function canMigrateConditionData(string $objectType): bool
+    {
+        return $this->migrateConditionObjectType !== null && $objectType === $this->migrateConditionObjectType;
```
```suggestion
        return $objectType === $this->migrateConditionObjectType;
```

`$objectType` can never be `null` thus a strict equality check will always fail when `$this->migrateConditionObjectType` is `null`.

Also, why does `null` prevent a migration here but allows it for `AbstractUserBooleanConditionType`?

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserStringConditionType.class.php:
```diff
@@ -103,4 +106,31 @@ private function getConditions(): array
+    #[\Override]
+    public function canMigrateConditionData(string $objectType): bool
+    {
+        return $this->migrateConditionObjectType !== null && $objectType === $this->migrateConditionObjectType;
```
See `AbstractUserBooleanConditionType`.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserStringConditionType.class.php:
```diff
@@ -103,4 +106,31 @@ private function getConditions(): array
+            return [];
+        }
+
+        $value = $conditionData[$this->migrateKeyName];
```
See `AbstractUserBooleanConditionType`.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/UserRegistrationDateConditionType.class.php:
```diff
@@ -85,4 +86,49 @@ private function getConditions(): array
+    private function convertDateStringTimestamp(string $date, int $hour, int $minute, int $seconds): int
+    {
+        $dateTime = new \DateTime($date, new \DateTimeZone(TIMEZONE));
+        $dateTime->setTime($hour, $minute, $seconds);
```
Please never use `\DateTime` and pick `\DateTimeImmutable` instead, otherwise this summons ~Cthulhu~ _Tim_.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/UserRegistrationDaysConditionType.class.php:
```diff
@@ -3,56 +3,32 @@
-        return 'registrationDays';
-    }
+        $container = parent::getFormField($id);
+        $filed = $container->getField();
```
Typo: `$field`.

dtdesign commented on wcfsetup/install/files/lib/system/user/group/assignment/command/UserGroupAssignmentMigrateCondition.class.php:
```diff
@@ -0,0 +1,42 @@
+
+        $migratedData = ConditionHandler::getInstance()->migrateConditionData(
+            new UserConditionProvider(),
+            JSON::decode($this->assignment->conditions)
```
This can yield an exception which will brick the process.

The deserialization should be a separate step and if it fails, this condition must be discarded because its broken anyway.

Cyperghost committed:
Rename 'needMigration' to 'isLegacy'

Cyperghost committed:
Apply suggestions from code review

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Merge remote-tracking branch 'origin/6.3-condition-migrate' into 6.3-condition-migrate

Cyperghost committed:
Refactor `ConditionMigration`

Cyperghost committed:
Refactor migration condition checks for consistency

Cyperghost committed:
Refactor condition data migration logic to improve null handling

Cyperghost committed:
Migrate user condition types to use named parameters for improved readability

Cyperghost committed:
Add legacy notice for automatic user group assignments and implement migration worker

Cyperghost committed:
Rename `UserGroupAssignmentMigrateCondition` to `MigrateLegacyCondition`

Cyperghost committed:
Fix typo in `$field`

Cyperghost committed:
Add error handling for JSON decoding in condition migration

Cyperghost committed:
Fix incorrect usage of unset function in ConditionHandler

Cyperghost committed:
Use `DateTimeImmutable` instead of `\DateTime`

Cyperghost committed:
Expand the PHPDoc for methods in `IMigrateConditionType`

dtdesign commented on wcfsetup/install/lang/de.xml:
```diff
@@ -870,6 +870,7 @@ Sie erreichen das Fehlerprotokoll unter: {link controller='ExceptionLogView' isE
 		<item name="wcf.acp.group.assignment.userGroup"><![CDATA[Benutzergruppe]]></item>
 		<item name="wcf.acp.group.assignment.isDisabled"><![CDATA[Automatische Zuordnung deaktivieren]]></item>
 		<item name="wcf.acp.group.assignment.title.error.tooLong"><![CDATA[Der eingegebene Titel ist zu lang. Die maximale Länge beträgt 255 Zeichen.]]></item>
+		<item name="wcf.acp.group.assignment.legacyNotice"><![CDATA[Führen Sie für die Automatische Benutzergruppen-Zuordnung die <a href="{link controller='RebuildData'}{/link}">Anzeigen aktualisieren</a> durch.]]></item>
```
```suggestion
		<item name="wcf.acp.group.assignment.legacyNotice"><![CDATA[Die automatischen Benutzergruppen-Zuordnungen wurden noch nicht aktualisiert. Bitte führen Sie dies unter <a href="{link controller='RebuildData'}{/link}">Anzeigen aktualisieren</a> durch.]]></item>
```

dtdesign commented on wcfsetup/install/lang/en.xml:
```diff
@@ -846,6 +846,7 @@ You can access the error log at: {link controller='ExceptionLogView' isEmail=tru
 		<item name="wcf.acp.group.assignment.userGroup"><![CDATA[User Group]]></item>
 		<item name="wcf.acp.group.assignment.isDisabled"><![CDATA[Disable automatic assignment]]></item>
 		<item name="wcf.acp.group.assignment.title.error.tooLong"><![CDATA[The entered title is too long. The maximum length is 255 characters.]]></item>
+		<item name="wcf.acp.group.assignment.legacyNotice"><![CDATA[For Automatic User Group Assignments, perform <a href="{link controller='RebuildData'}{/link}">Rebuild Data</a>.]]></item>
```
```suggestion
		<item name="wcf.acp.group.assignment.legacyNotice"><![CDATA[The automated user group assignments need to be recalculated. Please perform this task by visiting  <a href="{link controller='RebuildData'}{/link}">Rebuild Data</a>.]]></item>
```

dtdesign reviewed (changes requested):
Please also add an ACP status box of type warning to notify about outstanding migrations.

Cyperghost committed:
Apply suggestions from code review

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Add a status box if content needs to be recalculated using Rebuild Data.

dtdesign commented on wcfsetup/install/files/lib/system/acp/dashboard/box/StatusMessageAcpDashboardBox.class.php:
```diff
@@ -271,4 +273,42 @@ private function getPHPExtensionMessage(): array
+            $event->migrationNeeded(WCF::getLanguage()->get('wcf.acp.group.assignment'));
+        }
+
+        if ($event->needsMigration() !== []) {
+            return [
+                new StatusMessage(
+                    StatusMessageType::Warning,
+                    WCF::getLanguage()->getDynamicVariable('wcf.acp.dashboard.box.migrationNeeded', [
+                        'titles' => $event->needsMigration(),
+                    ])
+                ),
+            ];
+        }
+
+        return [];
```
Please invert this check, the happy path should always be at the bottom.

Cyperghost committed:
Invert check if migrations needed

dtdesign committed:
Avoid inheriting from concrete implementations

dtdesign committed:
Fix the class naming to be consistent

dtdesign committed:
`\strtolower()` is not binary-safe

dtdesign committed:
Fix the handling of condition types

dtdesign committed:
Mark the abstract user conditions as non-abstract

If we're going to instantiate them via anonymous classes, we can just drop the `abstract` because their entire implementation is non-abstract anyway.

dtdesign committed:
Register the conditions sequentually

There is very little gain in offering `addConditions()` plus it requires us to validate an arbitrary array in contrast to relying on `addCondition()` with its type hin.

dtdesign committed:
Fix the direction of the comparison

dtdesign committed:
Simplify the condition names

Avoiding the repetition makes it easier to find conditions.

Cyperghost committed:
Rename the class to the correct name
Add phpstan docs

Cyperghost committed:
Adding language variable
The content of `wcf.condition.add` has been swapped between German and English.
